### PR TITLE
allowing for graceful unicorn restarts

### PR DESCRIPTION
**Pull request by ghost**
This allows the web server to be gracefully restarted.

```
bin/eye restart web
```

### CODE REVIEW

**Comment by dmorley** [comment]
Nice! eye restart web worked perfectly with this for me.

**Comment by jramos** [comment]
Excellent. It momentarily complained that web had crashed for me on one restart, but the site seemed to continue working fine.

**Review by SuperTux88 — CHANGES_REQUESTED** [review]

**Comment by SuperTux88** [comment]
It also complains that it "crashes" here, but it seams to work ...

```
I, [2016-11-25T00:40:45.264533 #26825]  INFO -- : [Eye] => command: restart web
I, [2016-11-25T00:40:45.267179 #26825]  INFO -- : [diaspora:web] schedule :restart (reason: restart by user)
I, [2016-11-25T00:40:45.267536 #26825]  INFO -- : [diaspora:web] => restart  (reason: restart by user)
I, [2016-11-25T00:40:45.267768 #26825]  INFO -- : [diaspora:web] switch :restarting [:up => :restarting] (reason: restart by user)
I, [2016-11-25T00:40:45.289173 #26825]  INFO -- : [Eye] <= command: restart web (0.024521282s)
I, [2016-11-25T00:40:45.289531 #26825]  INFO -- : [diaspora:web] executing: `kill -USR2 27027` with restart_timeout: 10.0s and restart_grace: 1.0s
I, [2016-11-25T00:40:45.291858 #26825]  INFO -- : [diaspora:web] sleeping for :restart_grace 1.0
I, [2016-11-25T00:40:46.292451 #26825]  INFO -- : [diaspora:web] switch :restarted [:restarting => :up] (reason: restart by user)
I, [2016-11-25T00:40:46.302806 #26825]  INFO -- : [diaspora:web] <= restart
W, [2016-11-25T00:40:51.292904 #26825]  WARN -- : [diaspora:web] check_alive: pid_file (tmp/pids/web.pid) changed by itself (<27027> => <27285>), ignoring self-managed pid change
W, [2016-11-25T00:40:56.293151 #26825]  WARN -- : [diaspora:web] check_alive: process <27027> not found
I, [2016-11-25T00:40:56.293596 #26825]  INFO -- : [diaspora:web] switch :crashed [:up => :down] (reason: crashed)
I, [2016-11-25T00:40:56.294818 #26825]  INFO -- : [diaspora:web] schedule :check_crash (reason: crashed)
I, [2016-11-25T00:40:56.295324 #26825]  INFO -- : [diaspora:web] => check_crash  (reason: crashed)
W, [2016-11-25T00:40:56.295393 #26825]  WARN -- : [diaspora:web] check crashed: process is down
I, [2016-11-25T00:40:56.295456 #26825]  INFO -- : [diaspora:web] schedule :restore (reason: crashed)
I, [2016-11-25T00:40:56.295678 #26825]  INFO -- : [diaspora:web] <= check_crash
I, [2016-11-25T00:40:56.295819 #26825]  INFO -- : [diaspora:web] => restore  (reason: crashed)
I, [2016-11-25T00:40:56.296129 #26825]  INFO -- : [diaspora:web] load_external_pid_file: process <27285> from pid_file found and running (identity: ok) (unicorn master -c config/unicorn.rb -D)
I, [2016-11-25T00:40:56.296282 #26825]  INFO -- : [diaspora:web] switch :already_running [:down => :up] (reason: crashed)
I, [2016-11-25T00:40:56.308173 #26825]  INFO -- : [diaspora:web] <= restore
```

File: config/defaults.yml [hunk 1]
```diff
@@ -42,7 +42,7 @@ defaults:
   server:
     listen: '0.0.0.0:3000'
     rails_environment: 'development'
-    pid:
+    pid: "tmp/pids/web.pid"
```
Comment (jramos) [inline config/defaults.yml]: Done.

**Comment by jramos** [comment]
I think I fixed the "crash" issue by increasing the `restart_grace`. Let me know if it works for you...

File: config/unicorn.rb [hunk 2]
```diff
@@ -16,6 +16,16 @@
 before_fork do |_server, _worker|
   ActiveRecord::Base.connection.disconnect! # preloading app in master, so reconnect to DB
 
+  old_pid = "#{AppConfig.server.pid.get}.oldbin"
+
+  if File.exist?(old_pid) && _server.pid != old_pid
```
Comment (SuperTux88) [inline config/unicorn.rb]: Oh, please also change `_server` to `server`.

File: config/unicorn.rb [hunk 3]
```diff
@@ -13,9 +13,19 @@
 preload_app true
 @sidekiq_pid = nil
 
-before_fork do |_server, _worker|
+before_fork do |server, worker|
```
Comment (diaspora-code-review) [inline config/unicorn.rb]: Unused block argument - `worker`. If it's necessary, use `_` or `_worker` as an argument name to indicate that it won't be used.

File: config/unicorn.rb [hunk 4]
```diff
@@ -26,7 +36,7 @@
   end
 end
 
-after_fork do |_server, _worker|
+after_fork do |server, worker|
```
Comment (diaspora-code-review) [inline config/unicorn.rb]: - Unused block argument - `server`. You can omit all the arguments if you don't care about them.
- Unused block argument - `worker`. You can omit all the arguments if you don't care about them.

File: config/unicorn.rb (hunk 3 above)
Comment (svbergerem) [inline config/unicorn.rb]: https://wiki.diasporafoundation.org/Styleguide#Perform_an_automatic_local_review_before_pushing_to_your_repo

File: config/unicorn.rb [hunk 5]
```diff
@@ -26,7 +36,7 @@
   end
 end
 
-after_fork do |_server, _worker|
+after_fork do
```
Comment (SuperTux88) [inline config/unicorn.rb]: Please revert this line: `after_fork=#<Proc:0x00000002308a48@config/unicorn.rb:39> has invalid arity: 0 (need 2) (ArgumentError)`

Comment (jramos) [inline config/unicorn.rb]: My bad.

**Comment by SuperTux88** [comment]
Looks good now.

But sorry, I didn't see this before: Now it doesn't log the output of unicorn to stdout in development-mode when running script/server (because unicorn daemonizes itself). Maybe we should do this only in production mode? (we don't need zero-downtime-restarts in development-mode)

**Comment by jramos** [comment]
That makes sense. Updated to only happen in production.

File: config/eye.rb [hunk 6]
```diff
@@ -12,10 +12,21 @@
   stderr "log/eye_processes_stderr.log"
 
   process :web do
-    start_command "bin/bundle exec unicorn -c config/unicorn.rb"
-    daemonize true
-    pid_file "tmp/pids/web.pid"
+    @unicorn_command = "bin/bundle exec unicorn -c config/unicorn.rb"
```
Comment (SuperTux88) [inline config/eye.rb]: Please use a local variable here (without @).

**Review by SuperTux88 — CHANGES_REQUESTED** [review]
Looks good to me, only one small thing.

**Comment by jramos** [comment]
I actually refactored it a bit. Instead of killing off the old master immediately, it now removes a worker from the old master each time a new one is forked. The last worker forked then kills off the old master.

**Review by SuperTux88 — APPROVED** [review]

**Comment by SuperTux88** [comment]
Merged as 3586fa6ae701fc3dcd7d33a6ef0e5f5c46bf1b67

Thanks :)

**Comment by Flaburgan** [comment]
@jramos it would be awesome if you could also complete the wiki with that, for example in the FAQ for podmins ;)

**Comment by jramos** [comment]
@Flaburgan done: https://wiki.diasporafoundation.org/FAQ_for_pod_maintainers#How_can_I_restart_diaspora.2A_without_any_downtime.3F

**Comment by SuperTux88** [comment]
Maybe we should also add a note (at least until it is released and most of the pods upgraded), that this is only "without downtime" with 0.6.2.0 or newer. Otherwise it restarts with downtime.

**Comment by jramos** [comment]
@SuperTux88 updated.

**Pull request closed without merging**